packages/alfa-rules/src/sia-r65/rule.ts

```diff
@@ -25,10 +28,53 @@ export default Rule.Atomic.of<Page, Element, Question.Metadata>({
   requirements: [Criterion.of("2.4.7")],
   tags: [Scope.Component],
   evaluate({ device, document }) {
+    let diagnostic = Map.empty<string, Matches>();
+
+    const tabbables = document.tabOrder().filter(isTabbable(device));
+    const nonTargets = document
+      .descendants()
+      .filter((node) => isElement(node) && !tabbables.includes(node));
+    const targetClassnames = tabbables.flatMap((tabbable) => tabbable.classes);
+
+    let nonTargetClassnames = Sequence.empty<string>();
+    for (const nonTarget of nonTargets) {
+      if (isElement(nonTarget)) {
+        nonTargetClassnames = nonTargetClassnames.concat(nonTarget.classes);
+      }
+    }
+
+    function setMatches(
+      classname: string,
+      isTarget: boolean
+    ): Map<string, Matches> {
+      const matches: Matches = diagnostic
+        .get(classname)
+        .getOr({ matchingTargets: 0, matchingNonTargets: 0 });
+
+      const matchingTargets = isTarget
+        ? matches.matchingTargets + 1
+        : matches.matchingTargets;
+
+      const matchingNonTargets = isTarget
+        ? matches.matchingNonTargets
+        : matches.matchingNonTargets + 1;
```

Comment:
I feel this counting logic could be located with the data definition.

* `Matches` could be a full class with methods to increase number of targets or non targets, then we wouldn't need the logic to redefine new constants here.
* There could even be a wrapper class for `Map` that would  have a `addTarget(class: string)` method or something like that. Maybe it's easier to maintain two maps rather than a map whose values are pairs (OTOH, if it's consume as pair, it might be better to provide it that way).

For gathering class names, I'm also unsure about the logic. When computing `targetClassnames` (and `nonTargetClassnames`), we could directly reduce and crunch the numbers to avoid a bunch of `+1` later. (our `Sequence.groupBy` should do the trick). Then we can directly set the correct value. Here also, having a wrapping class can ease the logic with a `.from` function that does the needed (un)packing. So something like `targetClassnames.groupBy(class => class).map(list => list.size)` should directly provide a `Map` from class names to count.

(another way would be to get rid of the `targetClassnames` array and do a single pass through the DOM, adding classes to the diagnostic on the go. Something like `document.descendants().filter(isElement).forEach(element => { const isTarget = tabbales.includes(element); element.classes.forEach(class => matches.addCount(class, isTarget))})` with the `addCount` logic colocated with the data definition. While the call site looks elegant, I'm less sure about the side effects this uses 🤔 )

Reply:
To me it sounds like the easiest solution is to keep 2 maps that would then be serialized to:
```JSON
"targetClassname" : [
  [ "useless", 1 ],
  [ "good-focus", 2 ],
  [ "btn", 1 ],
  [ "irrelevant", 1 ]
],
"nonTargetClassname": [ [ "good-focus", 1 ], [ "irrelevant", 1 ] ]
```
TBH, I don't know how it will be consumed - if we need to look up how many target and non-target occurrences share what classes this might just do.

Reply:
Moving mapping of separate dictionaries to Dory